### PR TITLE
perf(sampler): draw the random bonus token in the rejection graph

### PR DESCRIPTION
**Pull request by rebel-eunik**
### 🚀 Summary of Changes

Follow-up to #1077, which moved the all-greedy bonus token into the rejection-sample graph.

#### Performance

**The random step's bonus token is drawn inside the rejection graph.**
Unless logprobs are requested or an argmax-invariant processor (min-p) is active, the graph receives the bonus rows' logits and a per-row temperature and draws with `rbln::top_k_top_p` under the same per-request `top_k`/`top_p` as the rejection op. A greedy row of a mixed batch carries `top_k == 1` and still argmaxes. The separate bonus sampler launch is gone for these steps.

**Logits rows are gathered with `torch.index_select`, and the postprocess gather becomes a view when the sampled rows lead the batch.**
`index_select` is the cheaper gather on the device. Decode+spec pads every request to K+1 slots, so when every request but the last carries the full draft count the sampled rows are a prefix of the logits and no gather is needed.

#### Bug fixes

- Warm-up covers every rejection graph a step can ask for.
- The rejection graph gets its own `top_k`/`top_p` buffers.
  the sampler read sampling metadata's tensors stale and emitted token id 0 for every row but the first under device tensors with random sampling, spec decode and `max_num_seqs > 1`. The impl now copies `top_k`/`top_p` into its own buffers

---

### 📌 Related Issues / Tickets

* Related to #1077
* Related to [fsw-efficiency#27](https://github.com/rebellions-sw/fsw-efficiency/issues/27)

---

### ✅ Type of Change

* [ ] 🚀 Release (`release`)
* [ ] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [x] 🛠 Bug fix (`fix`)
* [x] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

See [fsw-efficiency#27's comment](https://github.com/rebellions-sw/fsw-efficiency/issues/27#issuecomment-5659990031). 
When using n-gram (k=3), ITL improved by approximately 7-10% for Llama-3.2-1B and about 2% for MiniMax-M2.7 DP4EP

---

### 📋 Checklist

* [x] PR title follows Conventional Commits format
* [x] This PR is linked to an existing issue
* [x] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

**Not covered.**
Logprobs requests still take the bonus sampler by design, and a batch mixing greedy and random rows is exercised by unit tests only.

---

### CODE REVIEW

**Comment by rebel-daeyang** [comment]
@rebel-eunik A commit changing the KV cache layout landed on `dev`, so could you rebase once?

File: vllm_rbln/v1/worker/rbln_model_runner.py [hunk 1]
```diff
@@ -1857,7 +1857,12 @@ def execute_model(
             if not self.is_prefill and (
                 spec_decode_metadata is not None or self.uses_fixed_decode_window
             ):
-                logits = logits[logits_indices]
+                num_rows = logits_indices.shape[0]
+                if int(logits_indices[-1]) == num_rows - 1:
+                    # Ascending indices ending at num_rows - 1 are exactly 0..num_rows-1
+                    logits = logits[:num_rows]
+                else:
+                    logits = torch.index_select(logits, 0, logits_indices)
```
Comment (rebel-jonghewk) [inline vllm_rbln/v1/worker/rbln_model_runner.py:1867]: ~~@rebel-eunji~~ @rebel-eunik  Could you explain the intent behind this? I'd like to understand why this approach is faster.

Comment (rebel-eunik) [reply vllm_rbln/v1/worker/rbln_model_runner.py:1867]: If you were trying to tag me 😅,

A slice (`logits[:4]`) is a view sharing storage; indexing (`logits[[0,1,2,3]`) launches a gather kernel and allocates a new tensor. `logits_indices` is ascending (`query_start_loc[1:] - 1`), so when its last value is `num_rows - 1` it is exactly `0..num_rows-1` and the slice is equivalent; that is most spec-decode steps. 
Otherwise it gather with index_select, which traces showed to be cheaper on the device than normal indexing.

**Comment by rebel-eunik** [comment]
> @rebel-eunik A commit changing the KV cache layout landed on `dev`, so could you rebase once?

@rebel-daeyang  I've rebased onto dev(c529a99) and confirmed the performance improvement.

**Review by rebel-jonghewk — APPROVED** [review]
lgtm

**Review by rebel-cjlee — APPROVED** [review]

**Review by rebel-daeyang — APPROVED** [review]
